pkg/autodiscovery/providers/prometheus_services.go

```diff
@@ -20,42 +21,104 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apimachinery/pkg/labels"
 	listersv1 "k8s.io/client-go/listers/core/v1"
 	"k8s.io/client-go/tools/cache"
 )
 
+type ServiceAPI interface {
+	// List lists all Services
+	ListServices() ([]*v1.Service, error)
+	// GetEndpoints gets Endpoints by namespace and name
+	GetEndpoints(namespace, name string) (*v1.Endpoints, error)
+}
+
+type svcAPI struct {
+	serviceLister   listersv1.ServiceLister
+	endpointsLister listersv1.EndpointsLister
+}
+
+func (api *svcAPI) ListServices() ([]*v1.Service, error) {
+	return api.serviceLister.List(labels.Everything())
+}
+
+func (api *svcAPI) GetEndpoints(namespace, name string) (*v1.Endpoints, error) {
+	return api.endpointsLister.Endpoints(namespace).Get(name)
+}
+
 // PrometheusServicesConfigProvider implements the ConfigProvider interface for prometheus services
 type PrometheusServicesConfigProvider struct {
-	lister   listersv1.ServiceLister
+	sync.RWMutex
+
+	api      ServiceAPI
 	upToDate bool
-	PrometheusConfigProvider
```

Comment:
Sure. It is making testing very difficult because it depends on config and while we do have config mocking, it is still quite complex in practice. Instead ideally we should initialize checks (once?) and instantiate config providers with those checks, this can be done in the `New<ProviderType>` funcs. These changes are a step in the direction of simplifying dependencies.

Reply:
thanks, mind doing the same for the `prometheus_pods` providers?

Reply:
Sure sounds good.